astropy/timeseries/downsample.py

```diff
@@ -13,13 +13,35 @@
 __all__ = ["aggregate_downsample"]
 
 
+def nanmean_reduceat(data, indices):
+    mask = np.isnan(data)
+
+    if mask.any():  # If there are NaNs
+        # Create a writeable copy and mask NaNs
+        data = data.copy()
+        data[mask] = 0
+        count_data = np.add.reduceat(~mask, indices)
+        count_data = count_data.astype(float)
+        count_data[count_data == 0] = np.nan
+    else:
+        # Derive counts from indices
+        count_data = np.diff(indices, append=len(data))
+        count_data[count_data <= 0] = 1
+
+    sum_data = np.add.reduceat(data, indices)
+    nanmean = sum_data / count_data
+    return nanmean
+
+
 def reduceat(array, indices, function):
     """
     Manual reduceat functionality for cases where Numpy functions don't have a reduceat.
     It will check if the input function has a reduceat and call that if it does.
     """
     if len(indices) == 0:
         return np.array([])
+    elif function is nanmean_reduceat:
+        return np.array(function(array, indices))
```

Comment:
I may have missed a step, but this special casing should not be needed, right ? wasn't there a previous version of this PR were `nanmean_reduceat` was implemented such that it would naturally select the following branch without needing its own ?

Reply:
@neutrinoceros - this follows my recommendation to keep the implementation super simple and not add an attribute to a function.

That said, in this case I'm not sure at all why there is an `np.array` call here - that makes an unnecessary copy and really the output should be fine as is! (Indeed, it should be fine for the "reduceat" branch too. Though that perhaps should be for another PR...)

Reply:
In my opinion the complexity in branching here has greater implications.
`elif hasattr(function, "reduceat")` is already present and cannot be removed, obviously, so I'd rather leverage existing complexity than add special cases.

Reply:
The advantage of the current approach is that we're not "faking" a ufunc with a `reduceat` attribute, and that the standard path will be marginally faster.

It may also be the best path forward to allow people to pass in functions that have the right call signature - note that for things like taking a mean or median, `.reduceat` is really not a logical name, so I do not think we should be beholden to the hack that was put in place here and add another hack to make our default function work.

(Indeed, perhaps one change would be to call the function `nanmean_for_pieces` or so...)

Reply:
That sounds reasonable when you put it this way. Feel free to merge this as is if you're happy with it then ! :)